Added Decklist

XVicarious opened the pull request:
As you can see by the duplicated commit, I have no idea what I'm doing with git apparently.

Anyhow, I added the long talked about deck builder.

You can:  
* Make as many as you want  
* Have a sideboard, add to sideboard by long pressing add (is this intuitive?)  
* Change the card's set via click and dialog  
* Share a decklist  
* delete a decklist  

I don't know. The icon obviously needs to be changed, I used the about icon as a placeholder. I'm not artist.  Anything else? Exporting an analysis tools to come. I felt they weren't NEEDED upon release. I'll probably start them next week some time.

Addresses #49 .

XVicarious commented:
Okay, a couple of issues I've noticed since I submit this. I'll try to fix them tomorrow, Sunday at the latest.

* Current deck list isn't saved between changing fragments. 
* Changing device orientation clears the deck list

Both of these problems stem from the same place.

* Cards should also be sorted in WUBRG order, after CMC. 
* Missed a return statement on onOptionsItemSelected in DecklistFragment causing it to move onto the next menu

AEFeinstein commented:
It's gonna take me a while to review this, so be patient. I'll wait until your other fixes are in.
Also, you should add the fragment to default_fragment_array_entries and the widget buttonResources and intents.
I'll handle the iconography, don't worry about that.

XVicarious commented:
Everything should be fixed. I added common-collections for `ComparatorChain`. Support for this sort of thing sadly isn't in Android until API level 24, so an outside library had to be utilized.

XVicarious commented:
Fixed an issue I noticed today, and an oversight I made when converting the Comparators.

AEFeinstein commented:
I played around with the decklist fragment for a bit and I have some feedback.

Cards don't always get sorted into the right type, and Spear of Heliod seems to cause a number of problems.
- Cards with unknown supertypes are considered lands (Conspiracies, Schemes, Planes, Vanguards). They should go under an "Other" header
- Spells should be split into Instants and Sorceries.
- If you clear a deck and add Spear of Heliod, it doesn't get a type header.
- If you add Ordeal of Erebos, then Spear of Heliod, then Mishra's Bauble, all three cards get listed under Artifact.

Sometimes you can't modify the number of cards in a deck. To reproduce:
1. Clear app data
2. Add 1 "Vengeance"
3. Tap "Vengeance" and modify the quantity in the dialog
4. Notice the number in the deck does not change

The headings are tappable, which is fine, but it's a little odd that tapping on a heading selects the card below it.

The entries in decklist_menu.xml are missing this attribute: app:showAsAction="ifRoom"

The only deck I have saved is autosafe.fDeck (verified by adb shell), but there's an empty row below "autosave" in the deck selection dialog (load/delete)

All that said, it's a good start. Eventually when we build analysis tools in, I think it would work well in a new tab, like the Judge's Corner. That's just me brainstorming for now.

XVicarious commented:
Okay, patch incoming.

* Rewrote code for separators. The old code got confused about cards with multiple Super Types due to it checking the previous entry's types. It now checks the previous separator.  
* Inability to edit card numbers was due to an oversight in loading the deck name. If the deck you were editing was the autosave, it saved to ".fDeck", instead of "autosave.fDeck".  
* Multiple entries for the deck when only autosave existed was due to the previously mentioned bug.  
* .fDeck -> .deck, because the extension is just used internally.
* All menu entries now have the showAsAction set to ifRoom.  
* Added string entries for all card SuperTypes (currently "translatable=false", can be updated if multilanguage support for cards is added)  
* Added String Arrays for "card_types", "card_types_special", and a combined "card_type_extra".  
* Added String array for card type headers

As for the clickable headers. I don't know what to do about those. Currently even if "clickable" and "focusable" are set to false, the headers are clickable. Researching today, I did find another way. This would however require rewriting the adapter, because currently the headers are part of the ListView items themselves simply shown or hidden depending on if they're needed. This one is up to you.

XVicarious commented:
I just fixed a NullPointerException, but I'm also experiencing some odd bugs where there will be multiple headings for a type. Looking into it now.

AEFeinstein commented:
Okay. Let me know when you want me to look it over for pulling again.

On Mon, Mar 13, 2017, 7:00 PM Brian Maurer <notifications@github.com> wrote:

> I just fixed a NullPointerException, but I'm also experiencing some odd
> bugs where there will be multiple headings for a type. Looking into it now.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/AEFeinstein/mtg-familiar/pull/174#issuecomment-286270199>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOHDEm6wA_5zHrBbU9uTmNXMOhTVjNgks5rlcqBgaJpZM4MSSoz>
> .
>

XVicarious commented:
Okay, it should be good now.

AEFeinstein commented:
Adding "All Shall Smolder In My Wake" crashes the deck builder immediately, and will cause the deck builder to crash whenever it's opened. Same goes for the vanguard "Barrin." "Worldknit" works fine though. I recommend only keeping the list of sorted types (Creature, Instant, etc), and anything that doesn't match that list goes into "Other" (ditch card_types_special).

The divider in the menu drawer list is clickable, so I don't mind the clickable headers. I think the difference is, and I haven't looked at the source code, the card type headers are part of the first card's view while the menu drawer divider is its own list entry.

My only other request before merging the pull request is to display the total number of cards in the deck somewhere. A new row with the deck title and total number of cards would work, except a deck title isn't mandatory currently. What's your thought on prompting the user for a deck name if there are no saved decks via a dialog instead of using autosave.deck automatically?

Verified fixed:

- Spear of Heliod and Dryad Arbor are sorted correctly now.
- Instants and Sorceries are split.
- The number of Vengeances can be modified.
- The blank entry in the load/delete dialog is gone

XVicarious commented:
Damn, I swear I tested the other cards. Okay. Going to try to fix these today.

XVicarious commented:
Okay, I've fixed the crash and added a header.

https://imgur.com/a/LA1tc

What do you think about that? I also added a spot on the very right where a possible price might go.

The problem with adding separators like the drawer is that the deck list can be ever changing. It would require moving the entry around. It is possible, but I don't know how much trouble it would end up causing.

AEFeinstein commented:
FYI, if you save a deck, it doesn't update the deck name
